### PR TITLE
RDISCROWD-5733 Enforce size limit for task presenter HTML

### PR DESCRIPTION
**Pull request by kbecker42**
- Enforce size limit for task presenter HTML.
- Added check to view/projects.py (task presenter update from web page).
- Added check to api/project.py (task presenter update from PUT request).
- Size limit is configurable via settings_local.py value `TASK_PRESENTER_MAX_SIZE_MB`, default to 2 MB.
### Task presenter web page update

error-message-example

### Project API update

example-api-error

### CODE REVIEW

**Review by n00rsy — APPROVED** [review]
lgtm

File: pybossa/api/project.py [hunk 1]
```diff
@@ -51,6 +51,16 @@ class ProjectAPI(APIBase):
     private_keys = set(['secret_key'])
     restricted_keys = set()
 
+    def _preprocess_request(self, request):
+        # Limit maximum post data size.
+        content_length = request.content_length if request else None
```
[the remark this answers is not in the document]
Comment (dchhabda) [reply pybossa/api/project.py:56]: This code is computing size of the entire project payload and not just the task presenter code to be limited. Is this expected or it should only restrict size limit on task presenter code?

File: pybossa/api/api_base.py [hunk 2]
```diff
@@ -363,10 +364,15 @@ def _parse_request_data(self):
...
 
     def _preprocess_post_data(self, data):
-        """Method to be overriden by inheriting classes that will
+        """Method to be overridden by inheriting classes that will
         perform preprocessing on the POST data"""
         pass
 
+    def _preprocess_request(self, request):
+        """Method to be overridden by inheriting classes that will
+        perform preprocessong on the POST request"""
```
[the remark this answers is not in the document]
Comment (kbecker42) [reply pybossa/api/api_base.py]: Fixed via https://github.com/bloomberg/pybossa/pull/817/commits/72b951c61aef6280529ed88b08f78baafa7ca7e5

File: pybossa/api/project.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kbecker42) [reply pybossa/api/project.py]: Fixed via https://github.com/bloomberg/pybossa/pull/817/commits/72b951c61aef6280529ed88b08f78baafa7ca7e5

Indeed, the code will enforce the limit on HTML for the entire request (task presenter + guidelines). This intention is that internal systems would also be processing both, and thus, the need the limit the maximum size.

File: pybossa/view/projects.py [hunk 3]
```diff
@@ -606,13 +606,23 @@ def task_presenter_editor(short_name):
         flash(gettext(disabled_msg), 'error')
 
     if request.method == 'POST':
+        # Limit maximum post data size.
+        content_length = request.content_length
```
[the remark this answers is not in the document]
Comment (kbecker42) [reply pybossa/view/projects.py:610]: Fixed via https://github.com/bloomberg/pybossa/pull/817/commits/72b951c61aef6280529ed88b08f78baafa7ca7e5

File: pybossa/api/project.py [hunk 4]
```diff
@@ -51,6 +51,16 @@ class ProjectAPI(APIBase):
     private_keys = set(['secret_key'])
     restricted_keys = set()
 
+    def _preprocess_request(self, request):
+        # Limit maximum post data size.
+        content_length = request.content_length if request else 0
+        max_length_mb = current_app.config.get('TASK_PRESENTER_MAX_SIZE_MB', 2) if current_app and current_app.config else 0
```
Comment (dchhabda) [inline pybossa/api/project.py:57]: `max_length_mb = current_app.config.get('TASK_PRESENTER_MAX_SIZE_MB', 2)` seems sufficient to take care of missing configuration, and current_app would never be null

File: pybossa/api/project.py [hunk 5]
```diff
@@ -51,6 +51,16 @@ class ProjectAPI(APIBase):
...
+        max_length_bytes = max_length_mb * 1024 * 1024 # Maximum POST data size (MB)
+        if content_length and max_length_bytes and content_length > max_length_bytes:
```
Comment (dchhabda) [inline pybossa/api/project.py:59]: with proposed changes at line 57 ` if content_length > max_length_bytes:` would be sufficient and simplify if condition, hence no need to check `if content_length and max_length_bytes`

Comment (kbecker42) [reply pybossa/api/project.py]: Fixed via https://github.com/bloomberg/pybossa/pull/817/commits/227da7f1a58cfe1101b798586d008c5ea4ff23f0

File: pybossa/api/project.py (hunk 4 above)
Comment (kbecker42) [reply pybossa/api/project.py]: Fixed via https://github.com/bloomberg/pybossa/pull/817/commits/227da7f1a58cfe1101b798586d008c5ea4ff23f0

**Pull request merged**